### PR TITLE
MXDGA-3724: Added the MVP for the fast grid scan motion

### PR DESCRIPTION
**Pull request by DominicOram**
See https://jira.diamond.ac.uk/browse/MXDGA-3724

To review:
* Confirm that the Fast Grid Scan device is doing everything required for https://jira.diamond.ac.uk/browse/MXGDA-3733
* Confirm unit tests run
* Note that the system tests will not work against s03 currently as the fast grid scan IOC is not running on it

### CODE REVIEW

File: src/artemis/devices/unit_tests/test_gridscan.py
Comment (graeme-winter) [inline src/artemis/devices/unit_tests/test_gridscan.py:4]: Personally I prefer to import what you use rather than `from thing import *` - so for example when I wonder what `when()` is below, I can see where it comes from

File: src/artemis/devices/fast_grid_scan.py
Comment (graeme-winter) [inline src/artemis/devices/fast_grid_scan.py:103]: Kinda feel like more structured API here would help? `set_program_data` call in the test just looks like pushing a bunch of random numbers

**Review by graeme-winter — COMMENTED** [review]
Made a couple of comments, but these are just opinions on scanning the code. 

Do we use `black` for formatting?

I don't know what the "house style" is for blue sky so I don't know if `set_params(1, 2, 3, 4, 5)` is good practice or no...

File: src/artemis/devices/unit_tests/test_gridscan.py
Comment (DominicOram) [reply src/artemis/devices/unit_tests/test_gridscan.py:4]: Agreed, I wouldn't use `import *` in the actual production code so I shouldn't in the test code.

**Comment by DominicOram** [comment]
> Made a couple of comments, but these are just opinions on scanning the code.
> 
> Do we use `black` for formatting?
> 
> I don't know what the "house style" is for blue sky so I don't know if `set_params(1, 2, 3, 4, 5)` is good practice or no...

I've been using `black`, have a mild preference for it as it avoids arguments over specific formatting settings. Not sure if ophyd/bluesky use it. Looks like they use flake8 (https://github.com/bluesky/ophyd/blob/master/.flake8) but they're not necessarily mutually exclusive. Now that this repo is starting to grow I'll add a ticket in for some CI, including format checking against a similar flake8 profile to them.

File: src/artemis/devices/fast_grid_scan.py
Comment (DominicOram) [reply src/artemis/devices/fast_grid_scan.py:103]: Yh, I left this in as a bit of a placeholder until I had a better feel for how we would actually be using the Device upstream but I'll think about improving it now.

Comment (DominicOram) [reply src/artemis/devices/fast_grid_scan.py]: I borrowed some of the work of @callumforrester for this, which also included some checking against motor limits.

**Comment by graeme-winter** [comment]
> > Made a couple of comments, but these are just opinions on scanning the code.
> > Do we use `black` for formatting?
> > I don't know what the "house style" is for blue sky so I don't know if `set_params(1, 2, 3, 4, 5)` is good practice or no...
> 
> I've been using `black`, have a mild preference for it as it avoids arguments over specific formatting settings. Not sure if ophyd/bluesky use it. Looks like they use flake8 (https://github.com/bluesky/ophyd/blob/master/.flake8) but they're not necessarily mutually exclusive. Now that this repo is starting to grow I'll add a ticket in for some CI, including format checking against a similar flake8 profile to them.

We already have `black` and `flake8` set up as pre-commit hooks in most of our analysis code, so we can probably just copypasta the configuration over. Prefer pre-commit to working it out in a CI system as keeps the history cleaner... and makes for easier to read diffs.

File: src/artemis/devices/fast_grid_scan.py
Comment (callumforrester) [inline src/artemis/devices/fast_grid_scan.py:87]: It turns out there have been some issues with the `fraction` parameter in Ophyd. For some classes it goes from 0 -> 1, for other classes it goes the other way around. The correct behaviour is now documented here: https://github.com/bluesky/bluesky/pull/1496/files. I think this code needs to be reversed, unless I've read it wrong.

File: src/artemis/devices/fast_grid_scan.py
Comment (callumforrester) [inline src/artemis/devices/fast_grid_scan.py:91]: If there is some unknown exception I think we should pass it upwards with `self.set_exception()`.

File: src/artemis/devices/system_tests/test_gridscan_system.py
Comment (callumforrester) [inline src/artemis/devices/system_tests/test_gridscan_system.py:14]: I'm probably missing something obvious but I can't find this method, has it been renamed?

**Review by callumforrester — CHANGES_REQUESTED** [review]
Looks great, Dom. Code is very clear and well-documented. I particularly like the studious use of type annotations! Have added a few minor comments for things I spotted here and there.

**Comment by callumforrester** [comment]
@graeme-winter and @DominicOram 

> Prefer pre-commit to working it out in a CI system as keeps the history cleaner... and makes for easier to read diffs.

I prefer both because unknown agents can get around pre-commit, but pre-commit covers the 90% case so probably okay. However I suggest making a CI job for the tests regardless.

File: src/artemis/devices/fast_grid_scan.py
Comment (callumforrester) [reply src/artemis/devices/fast_grid_scan.py]: Didn't spot this chain, I think this needs updating to reflect this thread: https://github.com/DiamondLightSource/python-artemis/pull/2#discussion_r780119290

**Comment by thomascobb** [comment]
I'm going to put a plug in for how we do things in Controls, feel free to adopt whatever is useful:
- Pipenv for version management
- Pre-commit with black, flake8, isort and mypy for static analysis
- Pytest for code and coverage
- Sphinx for tutorials, how-to guides, explanations and reference documentation
- GitHub Actions for code and docs CI and deployment to PyPI and GitHub Pages
- VSCode settings using black, flake8, isort and mypy on save

None of this depends on internal DLS infrastructure so will still work if you unset the PYPI mirror.

We've captured this setup in a skeleton module that you can either copy/paste from or git pull changes from:
https://github.com/dls-controls/dls-python3-skeleton

Here's an example of a module that uses it:
https://github.com/dls-controls/sphinx_rtd_theme_github_versions

File: src/artemis/devices/motors.py
Comment (DominicOram) [reply src/artemis/devices/motors.py]: I agree that where I had it before is incorrect.

I've changed it to the responsibility of waiting lying with where the `MotorBundle` is created (e.g. see the tests). What do you think?

File: src/artemis/devices/fast_grid_scan.py
Comment (DominicOram) [reply src/artemis/devices/fast_grid_scan.py]: I used `ADTriggerStatus` (https://github.com/bluesky/ophyd/blob/master/ophyd/areadetector/trigger_mixins.py) as the basis for this, which I guess is also incorrect, I'll add a PR into `ophyd` to fix it.

**Comment by DominicOram** [comment]
Done the rework from both @callumforrester and @graeme-winter. CI issues will be addressed in https://jira.diamond.ac.uk/browse/MXGDA-3761

File: src/artemis/devices/fast_grid_scan.py
Comment (DominicOram) [reply src/artemis/devices/fast_grid_scan.py]: https://github.com/bluesky/ophyd/pull/1025 if you're interested

File: src/artemis/devices/motors.py
Comment (callumforrester) [reply src/artemis/devices/motors.py]: Looks good to me

**Comment by DominicOram** [comment]
After discussing with @graeme-winter today I think this is just waiting on review from @bentom08, if you have any thoughts?

**Review by bentom08 — APPROVED** [review]
Everything looks fine to me, happy for this to be merged in as is.

**Pull request merged**
